x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java

```diff
@@ -113,22 +117,20 @@ public boolean acquire(final EvictionListener listener) throws IOException {
         ensureOpen();
         boolean success = false;
         if (refCounter.tryIncRef()) {
-            evictionLock.lock();
             try {
-                ensureOpen();
-                final Set<EvictionListener> newListeners = new HashSet<>(listeners);
-                final boolean added = newListeners.add(listener);
-                assert added : "listener already exists " + listener;
-                maybeOpenFileChannel(newListeners);
-                listeners = Collections.unmodifiableSet(newListeners);
+                synchronized (listeners) {
+                    ensureOpen();
+                    final boolean added = listeners.add(listener);
```

Comment:
There's no need for a `volatile` and immutable set of listeners here. We never accessed the listeners outside of the lock so we can simply mutate them in place.
I made this clearer now by using the `listeners` set as the mutex for all operations that apply to the set and the channel opening/closing that is a result of the set size.